### PR TITLE
authentication via external services

### PR DESCRIPTION
**Pull request by DevMike**

### CODE REVIEW

File: app/controllers/registrations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class RegistrationsController < Devise::RegistrationsController
+  def update
+    # required for settings form to submit when password is left blank
+    if params[:user][:password].blank?
+      params[:user].delete("password")
+      params[:user].delete("password_confirmation")
+    end
+
+    @user = User.find(current_user.id)
```
Comment (kont-noor) [inline app/controllers/registrations_controller.rb:9]: why can't we use current_user?

File: app/models/user.rb [hunk 2]
```diff
@@ -1,26 +1,37 @@
+# encoding: utf-8
+
 class User < ActiveRecord::Base
   extend Enumerize
+
   CURRENCIES = %w(UAH RUB EUR USD)
   enumerize :preferred_currency, :in => CURRENCIES
+  OUTSIDE_AUTH_SERVICES = %w[facebook vkontakte]
```
Comment (kont-noor) [inline app/models/user.rb:8]: why [] not () ?

File: app/models/user.rb [hunk 3]
```diff
@@ -1,26 +1,37 @@
...
+  enumerize :provider, :in => OUTSIDE_AUTH_SERVICES
+  GENDERS = %w[автолюбитель автолюбительница]
```
Comment (kont-noor) [inline app/models/user.rb:10]: the same question

File: app/models/user.rb [hunk 4]
```diff
@@ -1,26 +1,37 @@
...
   belongs_to :settlement
 
   scope :confirmed, where("#{table_name}.confirmation_token IS NOT NULL")
   scope :unconfirmed, where(:confirmation_token => nil)
 
-  validates_presence_of :settlement_id, :name, :agree, :email
-  validates_presence_of :password, :password_confirmation, :on => :create
+  validates_presence_of :settlement_id, :name
+  validates_presence_of :password, :agree, :on => :create, :unless => :create_via_oauth?
+  validates_presence_of :password, :on => :update
```
Comment (kont-noor) [inline app/models/user.rb]: do we need this validation on update?

File: app/models/user.rb [hunk 5]
```diff
@@ -29,11 +40,65 @@ class User < ActiveRecord::Base
   
   has_attached_file :avatar, :styles => { :medium => "300x300>", :thumb => "100x100>" }
 
+  def full_name
+    "#{first_name} #{last_name}"
+  end
+
+  class << self
+    OUTSIDE_AUTH_SERVICES.each do |service|
+      define_method(:"create_or_find_by_#{service}_oauth") do |access_token|
```
Comment (kont-noor) [inline app/models/user.rb:49]: we can use general string as define_method agrument instead of symbol

File: app/models/user.rb [hunk 6]
```diff
@@ -29,11 +40,65 @@ class User < ActiveRecord::Base
...
+                       name: get_username(access_token, first_name, last_name),
+                       email: access_token.extra.raw_info.email,
+                       gender: gender,
+                       avatar: open(access_token.info.image))
+        end
+      end
+    end
+
+    def get_username(data, first_name, last_name)
+      name = [data.info.nickname, data.info.name, data.extra.raw_info.name, [first_name, last_name].compact.join(' ')].select(&:present?)[0]
```
Comment (kont-noor) [inline app/models/user.rb]: use detect instead of select ... [0]
detect returns single element

File: app/models/user.rb [hunk 7]
```diff
@@ -29,11 +40,65 @@ class User < ActiveRecord::Base
...
+      if exists?(name: name)
+        i = 0
+        while exists?(name: name) do
+          i += 1; name = "#{name}-#{i}"
+        end
+      end
```
Comment (kont-noor) [inline app/models/user.rb:79]: what about get user with (name like #{name}% sort by name desk group by name ?
and then increment name index. this should reduce queries abount

File: config/settings.yml [hunk 8]
```diff
@@ -15,3 +15,11 @@ global:
...
 mail:
   contact_mail: zarechenskiy.mihail@gmail.com #will be overwritten later
+
+auth:
+  facebook:
+    id: 118687028317192
+    secret: 814e59c1f27996c485c228bf10cb9650
+  vkontakte:
+    id: 3480385
+    secret: ZqnXnrSmwhlYNtcN3pkj
```
Comment (kont-noor) [inline config/settings.yml:25]: for now it's not critical but is it sensitive data?

File: app/controllers/registrations_controller.rb (hunk 1 above)
Comment (DevMike) [inline app/controllers/registrations_controller.rb:9]: it's migrated from devise, that allow update all users. I'm not sure that this "find" is needed for us, so it looks like your remark makes sense

File: app/models/user.rb (hunk 5 above)
Comment (DevMike) [inline app/models/user.rb:49]: what benefits of this way?

File: config/settings.yml (hunk 8 above)
Comment (DevMike) [inline config/settings.yml:25]: what do you mean exactly? These are credentials to our apps

File: app/models/user.rb (hunk 6 above)
Comment (DevMike) [inline app/models/user.rb]: done, thanks

File: app/models/user.rb (hunk 7 above)
Comment (DevMike) [inline app/models/user.rb:79]: used better way, see the commit

File: app/models/user.rb (hunk 5 above)
Comment (kont-noor) [inline app/models/user.rb:49]: just you shouldn't type colon

**Pull request merged**